airflow/models/taskinstance.py

```diff
@@ -2339,7 +2350,15 @@ def xcom_pull(
                 .order_by(None)
                 .order_by(XCom.map_index.asc())
             )
-            return (XCom.deserialize_value(r) for r in query)
+
+            def iter_xcom_values(query):
+                # The session passed to xcom_pull() may die before this is
+                # iterated through, so we need to bind to a new session.
+                with create_session() as session:
```

Comment:
why create a new one?

Reply:
Because `session` in the outer scope can die before we iterate through this, e.g.

```python
@task
def double(v):
    return v * 2

@task
def consume(v):
    resolved = list(v)  # The original session used to fetch XComs is dead by now.


consume(v=double.expand(v=[1, 2]))
```

Reply:
I'll take your word for it.  But that example seems quite odd... because xcom_pull isn't actually going to be invoked in what you've written, because no tasks have been run, so there's nothing to pull ...

Reply:
The return value from `double` tasks are pushed into XCom. This is similar to

```python
@task
def double(v):
    return v * 2

@task
def consume(v):
    print(v)

consume(v=double(v=2))
```

where `double` would push `4` to XCom.

Reply:
yeah i understand that, what i meant was, the example is just dag code, so the data movement is not happening in what you've written, right?  it only happens once you run the tasks.  i presume that to understand what's going with the sessions i need to understand more details about where / when / how this method is actually invoked. but i'll leave that for another day.